File: Heat.ConvertedToC#/Controllers/ProductInvoiceRowsController.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Xml.Linq;
using System.Diagnostics;
using System.Collections.Specialized;
using System.Configuration;
using System.Text;
using System.Text.RegularExpressions;
using System.Web;
using System.Web.Caching;
using System.Web.Mvc;
using System.Web.Mvc.Ajax;
using System.Web.Mvc.Html;
using System.Web.Routing;
using System.Web.SessionState;
using System.Web.Security;
using System.Web.Profile;
using System.Web.UI;
using System.Web.UI.WebControls;
using System.Web.UI.WebControls.WebParts;
using System.Web.UI.HtmlControls;
using System.Data.Entity;
using System.Net;
using Heat.Models;
using Heat.Repositories;
using Heat.ViewModels.Invoices;
using Heat.Manager;

namespace Heat.Controllers
{
	public class ProductInvoiceRowsController : System.Web.Mvc.Controller
	{

		private IHeatDBContext _db;
		private InvoiceModelBuilder _modelBuilder;

		private InvoiceManager _manager;

		public ProductInvoiceRowsController(IHeatDBContext context)
		{
			_db = context;
			_manager = new InvoiceManager(_db);
			_modelBuilder = new InvoiceModelBuilder(_db, _manager);

		}
		// GET: InvoiceRows
		public ActionResult Index()
		{
			return View(_db.ProductInvoiceRows.ToList());
		}

		// GET: InvoiceRows/Details/5
		public ActionResult Details(int? id)
		{
			if ((id == null)) {
				return new HttpStatusCodeResult(HttpStatusCode.BadRequest);
			}
			InvoiceRow invoiceRow = _db.InvoiceRows.Find(id);
			if ((invoiceRow == null)) {
				return HttpNotFound();
			}
			return View(invoiceRow);
		}

		[HttpGet()]
		public ActionResult Create(int invoiceID)
		{
			AddNewProductInvoiceRowViewModel invoiceRow = null;

			invoiceRow = _modelBuilder.GetAddProductInvoiceRowViewModel(invoiceID);

			return View(invoiceRow);
		}

		[HttpPost()]
		[ValidateAntiForgeryToken()]
		public ActionResult Create(AddNewProductInvoiceRowViewModel invoiceRow)
		{
			if (ModelState.IsValid) {
				ProductInvoiceRow invoiceRowDB = new ProductInvoiceRow();
				invoiceRowDB.Invoice = _db.Invoices.Find(invoiceRow.InvoiceID);
				if (_db.InvoiceRows.Where(x => x.Invoice.ID == invoiceRow.InvoiceID).Count() > 0) {
					invoiceRowDB.ItemOrder = _db.InvoiceRows.Where(x => x.Invoice.ID == invoiceRow.InvoiceID).Max(x => x.ItemOrder) + 1;
				} else {
					invoiceRowDB.ItemOrder = 1;
				}
				invoiceRowDB.Product = _db.Products.Find(invoiceRow.ProductID);
				invoiceRowDB.Quantity = invoiceRow.Quantity;
				invoiceRowDB.RateDiscount1 = (decimal) invoiceRow.Discount1;
				invoiceRowDB.RateDiscount2 = (decimal)invoiceRow.Discount2;
				invoiceRowDB.RateDiscount3 = (decimal)invoiceRow.Discount3;
				invoiceRowDB.UnitPrice = invoiceRow.UnitPrice;
				invoiceRowDB.VAT_Rate = invoiceRow.VAT;


				_db.ProductInvoiceRows.Add(invoiceRowDB);
				_db.SaveChanges();
				return RedirectToAction("edit", "invoices", new { ID = invoiceRow.InvoiceID });
			} else {
				ViewBag.message = "Errore nel salvataggio della riga";
				return View("error");
			}

		}


		public ActionResult Edit(int  id)
		{
			try {
				if ((id == null)) {
					return new HttpStatusCodeResult(HttpStatusCode.BadRequest);
				}

				//Dim invoiceRow As ProductInvoiceRow = _db.ProductInvoiceRows.Include(Function(x) x.Invoice).Include(Function(x) x.Product).Where(Function(x) x.ID = id).Single

				//If IsNothing(invoiceRow) Then
				//    Return HttpNotFound()
				//End If

				EditProductInvoiceRowViewModel model = new EditProductInvoiceRowViewModel();

				model = _modelBuilder.GetEditProductInvoiceRowViewModel(id);

				return View(model);
			} catch (Exception ex) {
				ViewBag.message = ex.Message;
				return View("error");
			}

		}


		[HttpPost()]
		[ValidateAntiForgeryToken()]
		public ActionResult Edit(EditProductInvoiceRowViewModel editedProductInvoiceRow)
		{
			if (ModelState.IsValid) {
				ProductInvoiceRow dbInvoiceRow = null;
				dbInvoiceRow = _db.ProductInvoiceRows.Find(editedProductInvoiceRow.ID);

				dbInvoiceRow.Product = _db.Products.Find(editedProductInvoiceRow.ProductID);
				dbInvoiceRow.Quantity = editedProductInvoiceRow.Quantity;
				dbInvoiceRow.RateDiscount1 = (decimal) editedProductInvoiceRow.Discount1;
				dbInvoiceRow.RateDiscount2 = (decimal) editedProductInvoiceRow.Discount2;
				dbInvoiceRow.RateDiscount3 = (decimal) editedProductInvoiceRow.Discount3;
				dbInvoiceRow.UnitPrice = editedProductInvoiceRow.UnitPrice;
				dbInvoiceRow.VAT_Rate = editedProductInvoiceRow.VAT;

				_db.SetModified(dbInvoiceRow);
				_db.SaveChanges();
				return RedirectToAction("edit", "invoices", new { ID = editedProductInvoiceRow.InvoiceID });
			} else {
				//il modello non è valido, torna alla vista di edit.
				return View(editedProductInvoiceRow);
			}

		}

		// GET: InvoiceRows/Delete/5
		public ActionResult Delete(int? id)
		{
			if ((id == null)) {
				return new HttpStatusCodeResult(HttpStatusCode.BadRequest);
			}
			InvoiceRow invoiceRow = _db.InvoiceRows.Find(id);
			if ((invoiceRow == null)) {
				return HttpNotFound();
			}
			return View(invoiceRow);
		}

		// POST: InvoiceRows/Delete/5
		[HttpPost()]
		[ActionName("Delete")]
		[ValidateAntiForgeryToken()]
		public ActionResult DeleteConfirmed(int id)
		{
			InvoiceRow invoiceRow = _db.InvoiceRows.Find(id);
			_db.InvoiceRows.Remove(invoiceRow);
			_db.SaveChanges();
			return RedirectToAction("Index");
		}

		protected override void Dispose(bool disposing)
		{
			if ((disposing)) {
				_db.Dispose();
			}
			base.Dispose(disposing);
		}
	}
}

File: Heat/Controllers/ProductInvoiceRowsController.vb
﻿Imports System
Imports System.Collections.Generic
Imports System.Data
Imports System.Data.Entity
Imports System.Linq
Imports System.Net
Imports System.Web
Imports System.Web.Mvc
Imports Heat.Models
Imports Heat.Repositories
Imports Heat.ViewModels.Invoices
Imports Heat.Manager

Namespace Controllers
    Public Class ProductInvoiceRowsController
        Inherits System.Web.Mvc.Controller

        Private _db As IHeatDBContext
        Private _modelBuilder As InvoiceModelBuilder
        Private _manager As InvoiceManager


        Sub New(context As IHeatDBContext)
            _db = context
            _manager = New InvoiceManager(_db)
            _modelBuilder = New InvoiceModelBuilder(_db, _manager)

        End Sub
        ' GET: InvoiceRows
        Function Index() As ActionResult
            Return View(_db.ProductInvoiceRows.ToList())
        End Function

        ' GET: InvoiceRows/Details/5
        Function Details(ByVal id As Integer?) As ActionResult
            If IsNothing(id) Then
                Return New HttpStatusCodeResult(HttpStatusCode.BadRequest)
            End If
            Dim invoiceRow As InvoiceRow = _db.InvoiceRows.Find(id)
            If IsNothing(invoiceRow) Then
                Return HttpNotFound()
            End If
            Return View(invoiceRow)
        End Function

        <HttpGet> _
        Function Create(invoiceID As Integer) As ActionResult
            Dim invoiceRow As AddNewProductInvoiceRowViewModel

            invoiceRow = _modelBuilder.GetAddProductInvoiceRowViewModel(invoiceID)

            Return View(invoiceRow)
        End Function

        <HttpPost()>
        <ValidateAntiForgeryToken()>
        Function Create(invoiceRow As AddNewProductInvoiceRowViewModel) As ActionResult
            If ModelState.IsValid Then
                Dim invoiceRowDB As New ProductInvoiceRow
                invoiceRowDB.Invoice = _db.Invoices.Find(invoiceRow.InvoiceID)
                If _db.InvoiceRows.Where(Function(x) x.Invoice.ID = invoiceRow.InvoiceID).Count > 0 Then
                    invoiceRowDB.ItemOrder = _db.InvoiceRows.Where(Function(x) x.Invoice.ID = invoiceRow.InvoiceID).Max(Function(x) x.ItemOrder) + 1
                Else
                    invoiceRowDB.ItemOrder = 1
                End If
                invoiceRowDB.Product = _db.Products.Find(invoiceRow.ProductID)
                invoiceRowDB.Quantity = invoiceRow.Quantity
                invoiceRowDB.RateDiscount1 = invoiceRow.Discount1
                invoiceRowDB.RateDiscount2 = invoiceRow.Discount2
                invoiceRowDB.RateDiscount3 = invoiceRow.Discount3
                invoiceRowDB.UnitPrice = invoiceRow.UnitPrice
                invoiceRowDB.VAT_Rate = invoiceRow.VAT


                _db.ProductInvoiceRows.Add(invoiceRowDB)
                _db.SaveChanges()
                Return RedirectToAction("edit", "invoices", New With {.ID = invoiceRow.InvoiceID})
            Else
                ViewBag.message = "Errore nel salvataggio della riga"
                Return View("error")
            End If

        End Function


        Function Edit(ByVal id As Integer?) As ActionResult
            Try
                If IsNothing(id) Then
                    Return New HttpStatusCodeResult(HttpStatusCode.BadRequest)
                End If

                'Dim invoiceRow As ProductInvoiceRow = _db.ProductInvoiceRows.Include(Function(x) x.Invoice).Include(Function(x) x.Product).Where(Function(x) x.ID = id).Single

                'If IsNothing(invoiceRow) Then
                '    Return HttpNotFound()
                'End If

                Dim model As New EditProductInvoiceRowViewModel

                model = _modelBuilder.GetEditProductInvoiceRowViewModel(id)

                Return View(model)
            Catch ex As Exception
                ViewBag.message = ex.Message
                Return View("error")
            End Try

        End Function

         
        <HttpPost()>
        <ValidateAntiForgeryToken()>
        Function Edit(ByVal editedProductInvoiceRow As EditProductInvoiceRowViewModel) As ActionResult
            If ModelState.IsValid Then
                Dim dbInvoiceRow As ProductInvoiceRow
                dbInvoiceRow = _db.ProductInvoiceRows.Find(editedProductInvoiceRow.ID)

                dbInvoiceRow.Product = _db.Products.Find(editedProductInvoiceRow.ProductID)
                dbInvoiceRow.Quantity = editedProductInvoiceRow.Quantity
                dbInvoiceRow.RateDiscount1 = editedProductInvoiceRow.Discount1
                dbInvoiceRow.RateDiscount2 = editedProductInvoiceRow.Discount2
                dbInvoiceRow.RateDiscount3 = editedProductInvoiceRow.Discount3
                dbInvoiceRow.UnitPrice = editedProductInvoiceRow.UnitPrice
                dbInvoiceRow.VAT_Rate = editedProductInvoiceRow.VAT

                _db.SetModified(dbInvoiceRow)
                _db.SaveChanges()
                Return RedirectToAction("edit", "invoices", New With {.ID = editedProductInvoiceRow.InvoiceID})
            Else
                'il modello non è valido, torna alla vista di edit.
                Return View(editedProductInvoiceRow)
            End If

        End Function

        ' GET: InvoiceRows/Delete/5
        Function Delete(ByVal id As Integer?) As ActionResult
            If IsNothing(id) Then
                Return New HttpStatusCodeResult(HttpStatusCode.BadRequest)
            End If
            Dim invoiceRow As InvoiceRow = _db.InvoiceRows.Find(id)
            If IsNothing(invoiceRow) Then
                Return HttpNotFound()
            End If
            Return View(invoiceRow)
        End Function

        ' POST: InvoiceRows/Delete/5
        <HttpPost()>
        <ActionName("Delete")>
        <ValidateAntiForgeryToken()>
        Function DeleteConfirmed(ByVal id As Integer) As ActionResult
            Dim invoiceRow As InvoiceRow = _db.InvoiceRows.Find(id)
            _db.InvoiceRows.Remove(invoiceRow)
            _db.SaveChanges()
            Return RedirectToAction("Index")
        End Function

        Protected Overrides Sub Dispose(ByVal disposing As Boolean)
            If (disposing) Then
                _db.Dispose()
            End If
            MyBase.Dispose(disposing)
        End Sub
    End Class
End Namespace
